### ISSUE TITLE
Best configuration for using ruby, rabbitmq and orientdb

### ISSUE THREAD

**Issue by AlexFiverr** [orientechnologies/orientdb#3850]
Hi, 

We have 10 rabbitmq workers updating orientdb. 
The workers receive updates on updating users, orders and conversations and updates the graph. 
each update assumes that migration has not run yet, and when we receive an update on an order, we assume the worst case - the users connected to the order and the order itself are not yet present in orient and thus we create them. 
we update the database via the REST API with a ruby gem - orientdb4r. 

the code looks something like this: 

``` ruby 
  def sync_order(order)
    return unless order
    buyer = User.find_by_id(order.user_id)
    seller = User.find_by_id(order.seller_id)
    
    json_for_buyer = format_and_jsonify(buyer, USER_WHITELIST_FIELDS)
    json_for_seller = format_and_jsonify(seller, USER_WHITELIST_FIELDS)
    json_for_order = format_and_jsonify(order, ORDER_WHITELIST_FIELDS)

    sync_objects = ["LET $buyer_vertex = UPDATE USER CONTENT #{json_for_buyer} UPSERT RETURN AFTER @rid WHERE id=#{buyer.id}",
                    "LET $seller_vertex = UPDATE USER CONTENT #{json_for_seller} UPSERT RETURN AFTER @rid WHERE id=#{seller.id}",
                    "LET $order = UPDATE USER CONTENT #{json_for_order} UPSERT RETURN AFTER @rid WHERE id=#{order.id}",
                    "DELETE EDGE bought FROM $buyer_vertex TO $order",
                    "DELETE EDGE sold FROM $seller_vertex TO $order",
                    "CREATE EDGE bought FROM $buyer_vertex TO $order",
                    "CREATE EDGE sold FROM $seller_vertex TO $order"]

    if order.status == 'completed'
      sync_objects.concat(["DELETE EDGE completed FROM $buyer_vertex TO $seller_vertex",
                           "CREATE EDGE completed_order FROM $buyer_vertex TO $seller_vertex"])
    end

    $orientdb.batch sync_objects

  end
``` 

the reason we are deleting and creating the edge is that orientdb does not have an upsert on edges yet. 

Running this, we get A LOT of errors - some legit - concurrency - we access the database via 10 workers, which sometimes get the same update at the same time - creating a race condition. we have the ``` commit retry 100``` code in our sql script we are sending in REST and we also run this as a transaction. 
But we get also a lot of errors like: 

Found a record (#0:0[0, 0, 0, 0] v1) that is not an edge. Record: #0:0[0, 0, 0, 0] v1 [OrientEdgeIterator]

or 

Record (#17:115515280) is null [OrientEdgeIterator]

or 

500 Internal Server Error: java.lang.NullPointerException

what are we doing wrong? I realise that these errors are all related to high concurrency, so my question is how to best configure the database to process these requests.

(running version 2.0.5) 

Thanks.

**Comment by lvca** [comment]
Please could you try with 2.0.6? @tglman did you remember this problem few weeks ago? Has been fixed?

**Comment by AlexFiverr** [comment]
Thanks @lvca ,We will try with 2.0.6, but can you please explain if those issues were fixed in this release? 
Also when using high speed concurrent updates, I find it hard to upsert edges...so if 2 workers got the same update at once, they may upsert the order which will result in one vertex, but when it comes to edges they may both add the "bought" edge, which will result in 2 different edges from the user to the order. 
I tried using in,out link properties in the edge scheme, with a unique index on them, but the exception being thrown on the duplicate entry fails the whole transaction. how can we do this in the most efficient and consistent manner? 

Thanks a lot

**Comment by AlexFiverr** [comment]
One more question: 
I returned to using in,out unique indexes on edges to keep only 1 edge between 2 vertices.
So when concurrently i try adding 2 edges I get:
```
com.orientechnologies.orient.core.storage.ORecordDuplicatedException: Cannot index record #17:532: found duplicated key 'OCompositeKey{keys=[#14:144679, #33:53938]}' in index 'participated.in_out' previously assigned to the record #17:500 RID=#17:500 [ONetworkProtocolHttpDb]
``` 

which is fine for me, but I wonder - the exception throwing on the server side - will it create a memory leak potentially? because i'm basically catching these exceptions and ignoring them, but it could create some load on the server...what do you think ?

**Comment by tglman** [comment]
hi,
this:
Found a record (#0:0[0, 0, 0, 0] v1) that is not an edge. Record: #0:0[0, 0, 0, 0] v1 [OrientEdgeIterator]

it's already fixed in 2.0.6.

this:
Record (#17:115515280) is null [OrientEdgeIterator]

it's not a problem and it doesn't create any real issue, is a corner case that can happen in high concurrency, we are already putting in place some solution to not let this happen in the first place, instead of handle after like now, probable 2.1 or 2.2 will have this fix.

this:
500 Internal Server Error: java.lang.NullPointerException

a null pointer is a problem, will be cool if we can have more information like a stack trace.

this:
```
com.orientechnologies.orient.core.storage.ORecordDuplicatedException: Cannot index record #17:532: found duplicated key 'OCompositeKey{keys=[#14:144679, #33:53938]}' in index 'participated.in_out' previously assigned to the record #17:500 RID=#17:500 [ONetworkProtocolHttpDb]
```
it not create any memory leak, and is not a huge overhead, i wouldn't worry much about that.

**Comment by AlexFiverr** [comment]
Hi @tglman , thanks a lot for your feedback. 

Regarding the ORecordDuplicatedException - we accepted this exception as it prevents creating duplicate edges between to vertices in high concurrency, BUT when we run in distributed mode, this exception fails the whole transaction and the replication fails: 

```
2015-04-02 14:19:28:410 SEVERE Internal server error:
com.orientechnologies.orient.core.exception.OTransactionException: Error on committing distributed transaction
--> com.orientechnologies.orient.core.storage.ORecordDuplicatedException: Cannot index record #17:2613: found duplicated key 'OCompositeKey{keys=[#14:1089, #12:2457]}' in index 'participated.in_out' previously assigned to the record #17:2178 RID=#17:2178 [ONetworkProtocolHttpDb]Error on distributed transaction commit
com.orientechnologies.orient.core.storage.ORecordDuplicatedException: Cannot index record #17:2613: found duplicated key 'OCompositeKey{keys=[#14:1089, #12:2456]}' in index 'participated.in_out' previously assigned to the record #17:2177 RID=#17:2177
	at com.orientechnologies.orient.core.index.OIndexUnique.commitSnapshot(OIndexUnique.java:173)
	at com.orientechnologies.orient.core.index.OIndexAbstract.commit(OIndexAbstract.java:704)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic$CommitIndexesCallback.run(OTransactionOptimistic.java:103)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.commit(OAbstractPaginatedStorage.java:922)
	at com.orientechnologies.orient.server.distributed.ODistributedStorage.commit(ODistributedStorage.java:811)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.doCommit(OTransactionOptimistic.java:483)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.commit(OTransactionOptimistic.java:147)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2391)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2361)
	at com.orientechnologies.orient.server.distributed.task.OTxTask.execute(OTxTask.java:113)
	at com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin.executeOnLocalNode(OHazelcastPlugin.java:757)
	at com.orientechnologies.orient.server.hazelcast.ODistributedWorker.onMessage(ODistributedWorker.java:300)
	at com.orientechnologies.orient.server.hazelcast.ODistributedWorker.run(ODistributedWorker.java:121)
```

is there any way we can suppress this exception and let the transaction finish?

**Comment by AlexFiverr** [comment]
Hi @tglman , @lvca , 

the exact exception is: 
```
2015-04-02 14:19:25:663 SEVERE Internal server error:
com.orientechnologies.orient.server.distributed.ODistributedException: Error on executing distributed request (id=9021 from=orientdb03 task=command_sql(DELETE EDGE completed_order FROM $buyer_vertex TO $seller_vertex) userName=) against database 'prod.[]' to nodes [orientdb04, orientdb03]
--> com.orientechnologies.orient.server.distributed.ODistributedException: Quorum 2 not reached for request (id=9021 from=orientdb03 task=command_sql(DELETE EDGE completed_order FROM $buyer_vertex TO $seller_vertex) userName=). Timeout=5ms Servers in timeout/conflict are:
 - orientdb03: com.orientechnologies.orient.core.exception.OCommandExecutionException: Error on execution of command: sql.DELETE EDGE completed_order FROM $buyer_vertex TO $seller_vertex
Received: {orientdb04=com.orientechnologies.orient.core.exception.OCommandExecutionException: Error on execution of command: sql.DELETE EDGE completed_order FROM $buyer_vertex TO $seller_vertex, orientdb03=com.orientechnologies.orient.core.exception.OCommandExecutionException: Error on execution of command: sql.DELETE EDGE completed_order FROM $buyer_vertex TO $seller_vertex} [ONetworkProtocolHttpDb]Error on distributed transaction commit
com.orientechnologies.orient.core.storage.ORecordDuplicatedException: Cannot index record #17:2605: found duplicated key 'OCompositeKey{keys=[#14:1002, #12:897]}' in index 'participated.in_out' previously assigned to the record #17:2003 RID=#17:2003
	at com.orientechnologies.orient.core.index.OIndexUnique.commitSnapshot(OIndexUnique.java:173)
	at com.orientechnologies.orient.core.index.OIndexAbstract.commit(OIndexAbstract.java:704)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic$CommitIndexesCallback.run(OTransactionOptimistic.java:103)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.commit(OAbstractPaginatedStorage.java:922)
	at com.orientechnologies.orient.server.distributed.ODistributedStorage.commit(ODistributedStorage.java:811)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.doCommit(OTransactionOptimistic.java:483)
	at com.orientechnologies.orient.core.tx.OTransactionOptimistic.commit(OTransactionOptimistic.java:147)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2391)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2361)
	at com.orientechnologies.orient.server.distributed.task.OTxTask.execute(OTxTask.java:113)
	at com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin.executeOnLocalNode(OHazelcastPlugin.java:757)
	at com.orientechnologies.orient.server.hazelcast.ODistributedWorker.onMessage(ODistributedWorker.java:300)
	at com.orientechnologies.orient.server.hazelcast.ODistributedWorker.run(ODistributedWorker.java:121)
```

any idea ? the duplication error on the index is important to us since it's the only way we can make sure double edges are not created in high concurrency...your help will be much appreciated!

**Comment by lvca** [comment]
Today I fixed a very similar issue in 2.0.8-SNAPSHOT. Please could you try it?

**Comment by prjhub** [comment]
The issue has been closed because no response has been received after a while. In case the issue is still valid, please reopen it or write a comment that explains why it's still valid and a way to reproduce it. Thank you.
